internal/server/data/grant.go

```diff
@@ -19,7 +40,26 @@ func CreateGrant(db GormTxn, grant *models.Grant) error {
 	case grant.Resource == "":
 		return fmt.Errorf("resource is required")
 	}
-	return add(db, grant)
+
+	// Use a savepoint so that we can query for the duplicate grant on conflict
+	if _, err := tx.Exec("SAVEPOINT beforeCreate"); err != nil {
+		// ignore "not in a transaction" error, because outside of a transaction
+		// the db conn can continue to be used after the conflict error.
+		if !isPgErrorCode(err, pgerrcode.NoActiveSQLTransaction) {
+			return err
+		}
+	}
+	if err := insert(tx, (*grantsTable)(grant)); err != nil {
+		_, _ = tx.Exec("ROLLBACK TO SAVEPOINT beforeCreate")
+		return handleError(err)
+	}
+	_, _ = tx.Exec("RELEASE SAVEPOINT beforeCreate")
```

Comment:
We weren't doing this previously. It's not strictly required, but I think it's a good idea so that we don't end up clobbering earlier savepoints with the same name.

I've also ignored the error for both the release and the rollback, since both of those operations will fail if we did not start the savepoint, and I think the error from the insert statement is going to be more relevant. Neither of these operations are required for functional correctness.